include/ensmallen_bits/pso/init_policies/feasible_init.hpp

```diff
@@ -0,0 +1,126 @@
+   *
+   * @param lowerBound Lower bound of the position initialization range.
+   * @param upperBound Upper bound of the position initialization range.
+   */
+  ConstrainedInit(const double lowerBound = -1.0,
+              const double upperBound = 1.0):
+      lowerBound(lowerBound),
+      upperBound(upperBound)
+  {
+    /* Nothing to do */
+  }
+
+  /**
+   * The InitializeParticles method of the init policy. Any class that is used
+   * in place of this default must implement this method which is used by the
+   * optimizer.
+   *
+   * @param iterate Coordinates of the initial point for training.
+   * @param numParticles The number of particles in the swarm.
+   * @param particlePositions Current positions of particles.
+   * @param particleVelocities Current velocities of particles.
+   * @param particleFitnesses Current fitness values of particles.
+   * @param particleBestPositions Best positions attained by each particle.
+   * @param particleBestFitnesses Best fitness values attained by each particle.
+   */
+  template<typename FunctionType>
+  void Initialize(FunctionType& function,
+			const arma::mat& iterate,
+                        const size_t numParticles,
+                        arma::cube& particlePositions,
+                        arma::cube& particleVelocities,
+                        arma::mat& particleFitnesses,
+                        arma::cube& particleBestPositions,
+                        arma::mat& particleBestFitnesses)
+  {
+    // Randomly initialize the particle positions.
+    particlePositions.randu(iterate.n_rows, iterate.n_cols, numParticles);
+    // User provided weights replacement performed here.
+    arma::mat positionVector;
+    double positionFitness;
+    double inf = std::numeric_limits<double>::max();
+    for (size_t i = 0; i < numParticles; i++)
+    {
+      bool positionFeasible = 0; //Flag for tracking if generated position vector is feasible or not
+      while (positionFeasible == 0)
+      {
+        //Generate random position vector
+        positionVector.randu(iterate.n_rows,iterate.n_cols);
+        positionFitness = function.Evaluate(positionVector);
+        if (positionFitness != inf)
+        {
+          particlePositions.slice(i) = positionVector;
+          positionFeasible = 1;
+        }
+      }
+    }
+
+    // Distribute particles in [lowerBound, upperBound].
+    particlePositions *= upperBound - lowerBound;
+    particlePositions += lowerBound;
+
+    // Randomly initialize particle velocities.
+    particleVelocities.randu(iterate.n_rows, iterate.n_cols, numParticles);
+
+    // Initialize current fitness values to infinity.
+    particleFitnesses.set_size(numParticles);
+    particleFitnesses.fill(std::numeric_limits<double>::max());
+
+    // Copy to personal best values for first iteration.
+    particleBestPositions = particlePositions;
+    // Initialize personal best fitness values to infinity.
+    particleBestFitnesses.set_size(numParticles);
```

Comment:
I guess if we like to parallelize the process at some point we need the fitness of each particle instead of just using an iterative process that saves the current best and worst fitness.

Reply:
When PSO is being run, there are essentially two loops. An _iteration_ in PSO would be an entire inner loop wherein the positions of all the particles are updated.

In order to stick to the original algorithm to the extent possible, what I want to do is implement parallelism on the updating _iteration_. You start a multi-thread or multi-process before iteration, and when the iteration completes, you wait for the multi-thread or multi-process to end. By doing so, we will be able to stick to the original algorithm without changes, but we will be able to make it faster by a constant amount of time.

Reply:
@zoq - does this sound alright? Would you be okay with this?

Reply:
Sounds good to me, I think the plan is to implement the standard method first, and afterwards implement parallelization.